L1Trigger/L1THGCal/src/be_algorithms/HGCalTowerMap2DImpl.cc

```
@@ -0,0 +1,155 @@
+///
+/// \class HGCalTowerMap2DImpl
+///
+/// \author: Thomas Strebler
+///
+/// Description: first iteration of HGCal Tower Maps
+
+#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTowerMap2DImpl.h"
+#include "DataFormats/Math/interface/normalizedPhi.h"
+
+constexpr double HGCalTowerMap2DImpl::kEtaMax_;
+constexpr double HGCalTowerMap2DImpl::kEtaMin_;
+constexpr double HGCalTowerMap2DImpl::kPhiMax_;
+constexpr double HGCalTowerMap2DImpl::kPhiMin_;
+
+
+HGCalTowerMap2DImpl::HGCalTowerMap2DImpl( const edm::ParameterSet& conf ) :
+  nEtaBins_(conf.getParameter<unsigned>("nEtaBins")),
+  nPhiBins_(conf.getParameter<unsigned>("nPhiBins")),
+  etaBins_(conf.getParameter<std::vector<double> >("etaBins")),
+  phiBins_(conf.getParameter<std::vector<double> >("phiBins")),
+  layerWeights_(conf.getParameter< std::vector<double> >("layerWeights"))
+{
+  edm::LogInfo("HGCalTowerMap2DImpl") << "Number of eta bins for the tower maps: " << nEtaBins_<<endl;
+  edm::LogInfo("HGCalTowerMap2DImpl") << "Number of phi bins for the tower maps: " << nPhiBins_<<endl;
+
+  if(etaBins_.size()>0 && etaBins_.size()!=nEtaBins_+1){
+    edm::LogError("HGCalTowerMap2DImpl") << "nEtaBins for the tower mapsnot consistent with etaBins size"<<endl;
```

We should maybe raise an exception. Otherwise unexpected behaviour may happen.